torch/csrc/jit/passes/quantization/freeze_quant_ops.cpp

```diff
@@ -0,0 +1,201 @@
+#include <stack>
+
+#include <torch/csrc/jit/api/module.h>
+#include <torch/csrc/jit/passes/constant_propagation.h>
+#include <torch/csrc/jit/passes/inliner.h>
+#include <torch/csrc/jit/passes/quantization/freeze_quant_ops.h>
+#include <torch/csrc/jit/runtime/graph_executor_impl.h>
+#include <torch/csrc/jit/passes/freeze_module.h>
+
+namespace torch {
+namespace jit {
+
+namespace {
+class QuantizationAttributeFreezer {
+ public:
+  QuantizationAttributeFreezer(
+      Module& module,
+      const FreezingOpsFilterFn& is_freezable_op)
+      : module_(module), is_freezable_op_(is_freezable_op) {
+    auto method = module.get_method("forward");
+    func_ = &method.function();
+  }
+
+  void run() {
+    auto graph = func_->graph();
+    // Inline the graph to make sure we don't have to traverse different
+    // subgraphs when trying to get the weight attributes.
+    Inline(*graph);
+    std::stack<Block*> blocks_to_visit({graph->block()});
+    while (!blocks_to_visit.empty()) {
+      Block* block = blocks_to_visit.top();
+      blocks_to_visit.pop();
+      for (auto n : block->nodes()) {
+        if (is_freezable_op_(n)) {
+          // Find getattrs in the input of the quant op
+          identifyAndConvertGetAttrs(module_, n, graph);
+        }
+        for (Block* sub_block : n->blocks()) {
+          blocks_to_visit.push(sub_block);
+        }
+      }
+    }
+    runOptimization(graph, false);
+    removeUnusedAttrs();
+    std::cout << "** FINAL Graph ** " << graph->toString();
+  }
+
+ private:
+  /*
+   * Find the GetAttr nodes in the graph that correspond to
+   * either the quantize prepack/run op or the
+   * quantize_per_tensor/quantize_per_channel operator
+   */
+  void identifyAndConvertGetAttrs(
+      Module& module,
+      Node* n,
+      std::shared_ptr<Graph>& graph) {
+    const auto& inputs = n->inputs().vec();
+    for (auto v : inputs) {
+      auto quant_inp_node = v->node();
+      WithInsertPoint ins(quant_inp_node);
+      if (quant_inp_node->kind() == prim::GetAttr) {
+        convertAttrToConst(module, quant_inp_node, graph);
+      } else if (
+          quant_inp_node->kind() == Symbol::aten("quantize_per_tensor") ||
+          quant_inp_node->kind() == Symbol::aten("quantize_per_channel")) {
+        identifyAndConvertGetAttrs(module, quant_inp_node, graph);
+      }
+    }
+  }
+
+  void convertAttrToConst(
+      Module& module,
+      Node* attr_node,
+      std::shared_ptr<Graph>& graph) {
+    // Map from module to list of constant attribute values in the graph
+    std::unordered_map<ModulePtr, std::unordered_map<std::string, Value*>>
+        moduleAttrs;
+    auto name = attr_node->s(attr::name);
+
+    auto attrModule = module;
+    std::deque<std::string> module_names;
+    // Get the module corresponding to the attr
+    if (!updateAttrModule(
```

Comment:
I think we have helper function for this? https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/quantization/helper.cpp.html#_ZN5torch3jit16getInvokedModuleERNS0_6ModuleEPNS0_4NodeEPNS0_5ValueE